### ISSUE TITLE
Allow owned types to have Id of base entity

### ISSUE THREAD

**Issue by 1234Georg** [dotnet/efcore#15884]
I would like to have the Ids of my entities in value objects. This would have several benefits in my model
- I get typed safety, as in the API methods of the system (e.g. Controller) I immediatly convert the id-strings to my value-objects an pass them further down
- these Id-Value-Objects  can have behavior (validate the id, make a new id after some rules)
- I can pass this objects arround without having to worry that they are exidentially changed

I tried various ways with owned types to make this happen. But none of them work as I would have expected it. For some I get an exception because the model is not correct (mainly because the id is not defined), or I get a QueryClientEvaluationWarning and the query is evaluated locally

Here is the domain model of a minimal solution, both Blog and Post have the id in a value object BlogId and PostId.

```
   public class Blog : IEquatable<Blog>
   {
        //für EF core
        private Blog() { }

        public Blog(BlogId blogId)
        {
            Id = blogId;
        }
        public BlogId Id { get; private set; }

        public bool Equals(Blog other)
        {
            return Id.Value == other?.Id.Value;
        }
    }

    public class Post : IEquatable<Post>
    {
        //für EF core
        private Post() { }

        public Post(PostId postId, BlogId blogId)
        {
            Id = postId;
            BlogId = blogId;
        }
        public BlogId BlogId { get; private set; }
        public PostId Id { get; private set; }

        public bool Equals(Post other)
        {
            return Id.Value == other?.Id.Value;
        }
    }

    public class BlogId : IEquatable<BlogId>
    {
        //für EF core
        private BlogId() { }
        public BlogId(string value)
        {
            Value = value;
        }
        public string Value { get; private set; }

        public bool Equals(BlogId other)
        {
            return Value == other.Value;
        }
    }

    public class PostId : IEquatable<PostId>
    {
        private PostId() { }
        public PostId(string value)
        {
            Value = value;
        }
        public string Value { get; private set; }

        public bool Equals(PostId other)
        {
            return Value == other.Value;
        }
    }
```

The following three options won't work because no Id is defined
```
            //with only the next line you get a exception "The entity type 'Blog' requires a primary key to be defined"
            //modelBuilder.Entity<Blog>().OwnsOne(p => p.Id);

            //with only the next line still the exception: The entity type 'Blog' requires a primary key to be defined
            //modelBuilder.Entity<Blog>().OwnsOne(b => b.Id, sb => sb.HasKey(id => id.Value));

            //with the next two lines there is the exception: "The properties expression 'b => b.Id.Value' is not valid. The expression should represent a simple property access: 't => t.MyProperty'. ..."
            //modelBuilder.Entity<Blog>().OwnsOne(b => b.Id);
            //modelBuilder.Entity<Blog>().HasKey(b => b.Id.Value);
```

The next option uses conversion. This would work, if I would allow client side evaluation. However this is not an option because it's way to slow in my situation. The other downside of this approach is, that I can't use combined keys, as the conversion takes only one parameter.
```
modelBuilder.Entity<Blog>().Property(b => b.Id).HasConversion(idObject => idObject.Value, idString => new BlogId(idString));
```

In all cases I try to query it like this, see HomeController.Index
```
var blog456 = DbContext.Blogs.Where(p => p.Id == new BlogId("456")).ToList();
```

What would really work for me, is some way to tell the entity that its key is in that owned type.

This is the QueryClientEvaluationWarning-Exception

```
Exception message:
Error generated for warning 'Microsoft.EntityFrameworkCore.Query.QueryClientEvaluationWarning: The LINQ expression 'where ([p].Id == value(EfCore30_IdsInValueObjects.BlogId))' could not be translated and will be evaluated locally.'. This exception can be suppressed or logged by passing event ID 'RelationalEventId.QueryClientEvaluationWarning' to the 'ConfigureWarnings' method in 'DbContext.OnConfiguring' or 'AddDbContext'.

Stack trace:
   at Microsoft.EntityFrameworkCore.Diagnostics.EventDefinition`1.Log[TLoggerCategory](IDiagnosticsLogger`1 logger, WarningBehavior warningBehavior, TParam arg)
```

### Steps to reproduce
I made a minimal solution, you can download it here: https://github.com/1234Georg/EfCore30_IdsInValueObjects.git
Just press F5
- the query is in the HomeController.cs
- the model is in DomainModel.cs
- DbContext is in BlogDbContext

### Further technical details
EF Core version: 3.0.0-preview5.19227.1
Database Provider: Microsoft.EntityFrameworkCore.Sqlite, Version 3.0.0-preview5.19227.1
Operating system: Windows 10
IDE: Visual Studio 2019 16.0.3

**Comment by ajcvickers** [comment]
@1234Georg A few comments:
* `Blog` and `Post` have no navigation properties between them. An owning entity type needs to have a reference to or collection of the owned types.
* Calling `OwnsOne` on the key property (instead of using a navigation property to the owned type) is an error. In essence, you are telling EF to make the key type an owned entity type.
* It's quite strange to have the `Blog` be owned by the `Post`. You might want to consider if this is really the right way to model ownership. In other words, is `Post` really the aggregate root for an aggregate that also contains the related `Blog`?

Can you either update the code to fix these issues, or if they are by-design can you provide some more details on why/how?

**Comment by 1234Georg** [comment]
@ajcvickers 
Thanks for your response.

- In the example I have two aggregates: Blog and Post
- They don't have a navigation property, Post references Blog just by the id of Blog: Post.BlogId, so it's a disconnected domain model, according to Vaughn Vernon.
- What I want to achieve: I want to have the Id of Blog and Post in a Value-Object (in DDD notation), in the example it's BlogId and PostId. To map these to the database I tried to use owned-types of ef core. But that would only work with HasConversion and client-side evaluation. And it will only works for single valued keys and not for combined keys. 

So these are my references
Blog    --->   BlogId
Post    --->   PostId
Post    --->   BlogId

The advantage of having the Ids in Value-Objects are, as I already wrote above:
- I get type safety, as in the API methods of the system (e.g. Controller) I immediatly convert the id-strings to my value-objects an pass them further down.
- these Id-Value-Objects can have behavior (validate the id, make a new id after some rules)
- I can pass these objects arround without having to worry that they are accidentially changed because they are immutable
- I would like to consider the Ids as part of my ubiquituous language

If modelling the Ids using owned types are not the way to go, I'm happy to get other suggestions.

**Comment by ajcvickers** [comment]
@1234Georg Thanks--that makes a lot more sense. This should only require value converters--there is no need to try to map the `Id` types.
```C#
modelBuilder.Entity<Blog>(b =>
{
    b.Property(b => b.Id).HasConversion(idObject => idObject.Value, idString => new BlogId(idString));
});

modelBuilder.Entity<Post>(b =>
{
    b.Property(b => b.Id).HasConversion(idObject => idObject.Value, idString => new PostId(idString));
    b.Property(b => b.BlogId).HasConversion(idObject => idObject.Value, idString => new BlogId(idString));
});
```

However, I think you are then hitting this limitation: #10434 for which there isn't currently any good workaround, but which is scheduled to be worked on for 3.0.

**Comment by 1234Georg** [comment]
@ajcvickers 
Thanks for your response. Good to hear that you plan to fix this in EF Core 3.0.
I made a follow-up issue on howto map combined keys in value-objects: #15962
